### ISSUE TITLE
4.9.0 esbuild plugin causes [Cannot read properties of undefined (reading 'inherit')] on lambda init

### ISSUE THREAD

**Issue by kcveitch** [DataDog/dd-trace-js#3479]
Testing 4.9.0 in our build process specifically to test #3448 which I believe to be the cause of some of our previous instrumentation issues. I made no changes to code and simply installed dd-trace@4.9.0 in the build pipeline.  Build completed successfully but the deployed lambda function immediately fails during initialization phase with the following error:

```
INIT_START Runtime Version: nodejs:16.v16	Runtime Version ARN: arn:aws:lambda:us-east-1::runtime:c1983f4e9ce3930d3093fa38128e12edac2324f4fe99adfb8e0d5fb6b6afadbc

TELEMETRY	Name: datadog-agent	State: Subscribed	Types: [Platform, Function, Extension]
2023-07-27T23:18:17.929Z	undefined	ERROR	Uncaught Exception 	
{
    "errorType": "TypeError",
    "errorMessage": "Cannot read properties of undefined (reading 'inherit')",
    "stack": [
        "TypeError: Cannot read properties of undefined (reading 'inherit')",
        "    at /var/task/app.js:1:45662",
        "    at /var/task/app.js:1:227",
        "    at /var/task/app.js:16:811",
        "    at /var/task/app.js:1:227",
        "    at /var/task/app.js:16:1124",
        "    at /var/task/app.js:1:227",
        "    at /var/task/app.js:16:22349",
        "    at /var/task/app.js:1:227",
        "    at /var/task/app.js:16:27993",
        "    at /var/task/app.js:1:227"
    ]
}

2023-07-27T23:18:17.934Z	undefined	DEBUG	Inject into carrier: 
{
    "x-datadog-trace-id": "3697252533875077936",
    "x-datadog-parent-id": "3697252533875077936",
    "x-datadog-sampling-priority": "1"
}

```

Reverting the build pipeline to use dd-trace@4.8.1 resolves the error and returns us to the previous semi functional state we had before in which the dd supported integrations are not instrumented.

**Steps to reproduce**

1. build TypeScript project using dd-trace esbuild plugin on version 4.9.0 and the package.json and esbuild.config.js defined below
2. resulting lambda may fail in a similar way
3. revert to dd-trace@4.8.1
4. confirm functional lambda again

**Environment**

* **Operation system: ** Amazon Linux 2
* **Node.js version:** 16
* **Tracer version:** 
* Build time: 4.9.0   
* Runtime: lambda nodejs layer version 95 (dd-trace-js ^3.29.1)
* **Agent version:** Lambda Extension Layer 45 (Agent: 7.46.0)

Package json Dependencies:

```
  "dependencies": {
    "dotenv": "^16.0.3",
    "esbuild": "^0.18.2",
    "mongoose": "^7.3.0"
  },
  "devDependencies": {
    "@aws-lambda-powertools/logger": "^1.9.0",
    "@aws-lambda-powertools/metrics": "^1.9.0",
    "@aws-lambda-powertools/tracer": "^1.9.0",
    "@types/aws-lambda": "^8.10.117",
    "@types/jest": "^27.4.0",
    "@types/node": "^20.3.1",
    "@typescript-eslint/eslint-plugin": "^5.10.2",
    "@typescript-eslint/parser": "^5.10.2",
    "aws-lambda": "^1.0.7",
    "aws-sdk": "^2.1399.0",
    "esbuild-jest": "^0.5.0",
    "eslint": "^8.8.0",
    "eslint-config-prettier": "^8.3.0",
    "eslint-plugin-prettier": "^4.0.0",
    "jest": "^29.5.0",
    "prettier": "^2.5.1",
    "ts-jest": "^29.1.0",
    "ts-node": "^10.9.1",
    "typescript": "^5.1.6"
  }

```

esbuild.config.js:

```
const ddPlugin = require('dd-trace/esbuild')
const { build } = require("esbuild");

console.log(`*************************** current directory: ${__dirname}`);
let dirN;

for (const x in process.argv) {
    const e = process.argv[x];
    console.log(e + " " + x);
    if (e == '--dir') {
        // get the dir value from the next argument
        const nextV = parseInt(x) + 1;
        dirN = process.argv[nextV];
    }
}

if (dirN == null) {
    dirN =  __dirname + "./.aws-sam/build/<function name redacted>/";
}

if (!dirN.endsWith("/")) {
    dirN = `${dirN}/`;
}

build({
    entryPoints: [`${__dirname}/src-ts/app.ts`],
    target: "node16",
    bundle: true,
    platform: "node",
    plugins: [ddPlugin],
    // external: ['graphql/language/visitor', 'graphql/language/printer', 'graphql/utilities', 'aws-sdk', 'datadog-lambda-js', 'dd-trace' ],
    sourcemap: true,
    minify: true,
    outfile: `${dirN}app.js`
}).catch((error) => {
    console.error(error);
    process.exit(1);
});

```

**Comment by tlhunter** [comment]
@kcveitch thanks for the bug report! I'm not as familiar with the Lambda side of things but I built the esbuild feature so I'll do my best.

First things to try: can you add `"dd-trace", "datadog-lambda-js", "aws-sdk"` to the `external` list? Those are provided by the lambda layer and shouldn't benefit your build.

Also, can you try setting `minify: false` and see if this changes anything?

Can you set `DD_TRACE_DEBUG=true` when running the build process? When set, there will be additional output that might be helpful to debug a bad build.

Also, when you run the lambda, can you set `DD_LOG_LEVEL=debug` and look for anything suspicious in the output?

The `inherit` property that is being accessed _somewhere_ in the application doesn't appear to be a property that we're referencing in dd-trace. So I wonder if the issue is with interplay between how Lambda is invoking the compiled function. It might also be a problem with how the esbuild plugin is interacting with one of your dependencies (even a deeply nested sub-dependency). That said your dependency list seems tiny so I'm not sure if that's a real concern. When I run a basic node app with mongo/mongoose locally I don't see an error like what you're getting.

**Comment by kcveitch** [comment]
Updated External modules in esbuild, minify:false, and DD_LOG_LEVEL:debug per your request.  see resulting logs:
```

INIT_START Runtime Version: nodejs:16.v16	Runtime Version ARN: arn:aws:lambda:us-east-1::runtime:c1983f4e9ce3930d3093fa38128e12edac2324f4fe99adfb8e0d5fb6b6afadbc
2023-07-31 15:57:19 UTC | DD_EXTENSION | DEBUG | Datadog extension version : 45|Datadog environment variables: DD_API_KEY=***|DD_CAPTURE_LAMBDA_PAYLOAD=true|DD_ENHANCED_METRICS=true|DD_ENV=stage|DD_FLUSH_TO_LOG=true|DD_LAMBDA_HANDLER=src-ts/app.lambdaHandler|DD_LOGS_INJECTION=true|DD_LOG_LEVEL=debug|DD_MERGE_XRAY_TRACES=false|DD_PROFILING_ENABLED=true|DD_SERVERLESS_LOGS_ENABLED=true|DD_SERVICE=<REDACTED NAME>|DD_SITE=datadoghq.com|DD_TAGS=git.commit.sha:"***********************************110dc",git.repository_url:https://bitbucket.org/<REDACTED ORG>/<REDACTED NAME>|DD_TRACE_DEBUG=true|DD_TRACE_ENABLED=true|DD_TRACE_STARTUP_LOGS=true|DD_VERSION=106|
2023-07-31 15:57:19 UTC | DD_EXTENSION | DEBUG | Starting daemon to receive messages from runtime...
2023-07-31 15:57:19 UTC | DD_EXTENSION | DEBUG | Unable to restore the state from file
2023-07-31 15:57:19 UTC | DD_EXTENSION | DEBUG | No config file detected, using environment variable based configuration only
2023-07-31 15:57:19 UTC | DD_EXTENSION | DEBUG | FIPS mode is disabled
2023-07-31 15:57:19 UTC | DD_EXTENSION | INFO | 0 Features detected from environment: 
2023-07-31 15:57:19 UTC | DD_EXTENSION | DEBUG | Using a SyncForwarder with a 5s timeout
2023-07-31 15:57:19 UTC | DD_EXTENSION | INFO | Retry queue storage on disk is disabled
2023-07-31 15:57:19 UTC | DD_EXTENSION | INFO | Creating TimeSampler #0
2023-07-31 15:57:19 UTC | DD_EXTENSION | DEBUG | 'telemetry.dogstatsd.aggregator_channel_latency_buckets' is empty, falling back to default values
2023-07-31 15:57:19 UTC | DD_EXTENSION | DEBUG | 'telemetry.dogstatsd.listeners_latency_buckets' is empty, falling back to default values
2023-07-31 15:57:19 UTC | DD_EXTENSION | DEBUG | 'telemetry.dogstatsd.listeners_channel_latency_buckets' is empty, falling back to default values
2023-07-31 15:57:19 UTC | DD_EXTENSION | DEBUG | Forwarder started
2023-07-31 15:57:19 UTC | DD_EXTENSION | DEBUG | Demultiplexer started
2023-07-31 15:57:19 UTC | DD_EXTENSION | DEBUG | dogstatsd-udp: 127.0.0.1:8125 successfully initialized
2023-07-31 15:57:19 UTC | DD_EXTENSION | DEBUG | DogStatsD will run 2 workers
2023-07-31 15:57:19 UTC | DD_EXTENSION | INFO | dogstatsd-udp: starting to listen on 127.0.0.1:8125
2023-07-31 15:57:19 UTC | DD_EXTENSION | DEBUG | No config file detected, using environment variable based configuration only
2023-07-31 15:57:19 UTC | DD_EXTENSION | DEBUG | Enabling telemetry collection HTTP route
2023-07-31 15:57:19 UTC | DD_EXTENSION | DEBUG | Subscribing to Telemetry for types: [platform function extension]
2023-07-31 15:57:19 UTC | DD_EXTENSION | DEBUG | appsec: security monitoring is not enabled: DD_SERVERLESS_APPSEC_ENABLED is not set to true
TELEMETRY	Name: datadog-agent	State: Subscribed	Types: [Platform, Function, Extension]
2023-07-31 15:57:19 UTC | DD_EXTENSION | DEBUG | FIPS mode is disabled
2023-07-31 15:57:19 UTC | DD_EXTENSION | INFO | 0 Features detected from environment: 
2023-07-31 15:57:19 UTC | DD_EXTENSION | DEBUG | otlp endpoint disabled
2023-07-31 15:57:19 UTC | DD_EXTENSION | INFO | Loaded configuration: /var/task/datadog.yaml
2023-07-31 15:57:19 UTC | DD_EXTENSION | DEBUG | Setting DefaultEnv to "stage" (from 'env' config option)
2023-07-31 15:57:19 UTC | DD_EXTENSION | INFO | Starting logs-agent...
2023-07-31 15:57:19 UTC | DD_EXTENSION | INFO | logs-agent started
2023-07-31 15:57:19 UTC | DD_EXTENSION | DEBUG | Adding AWS Logs Log Source
2023-07-31 15:57:19 UTC | DD_EXTENSION | DEBUG | Starting ColdStartSpanCreator
2023-07-31 15:57:19 UTC | DD_EXTENSION | DEBUG | Setting ColdStartSpanCreator on Daemon
2023-07-31 15:57:19 UTC | DD_EXTENSION | DEBUG | serverless agent ready in 27.263866ms
2023-07-31T15:57:20.478Z	undefined	ERROR	Uncaught Exception 	
{
    "errorType": "TypeError",
    "errorMessage": "Cannot read properties of undefined (reading 'inherit')",
    "stack": [
        "TypeError: Cannot read properties of undefined (reading 'inherit')",
        "    at node_modules/aws-sdk/lib/sequential_executor.js (/var/task/app.js:3154:41)",
        "    at __require (/var/task/app.js:9:50)",
        "    at node_modules/aws-sdk/lib/core.js (/var/task/app.js:9761:5)",
        "    at __require (/var/task/app.js:9:50)",
        "    at datadog:/tmp/tmpmpgcl__l/node_modules/aws-sdk/lib/core.js (/var/task/app.js:9784:15)",
        "    at __require (/var/task/app.js:9:50)",
        "    at node_modules/aws-sdk/lib/event-stream/event-message-chunker-stream.js (/var/task/app.js:11343:16)",
        "    at __require (/var/task/app.js:9:50)",
        "    at node_modules/aws-sdk/lib/event-stream/streaming-create-event-stream.js (/var/task/app.js:11739:37)",
        "    at __require (/var/task/app.js:9:50)"
    ]
}

2023-07-31T15:57:20.484Z	undefined	DEBUG	Inject into carrier: 
{
    "x-datadog-trace-id": "4706390504097108744",
    "x-datadog-parent-id": "4706390504097108744",
    "x-datadog-sampling-priority": "1"
}
.
2023-07-31T15:57:20.499Z	undefined	DEBUG	Encoding payload: [
{
    "trace_id": "4150753a5afad308",
    "span_id": "4150753a5afad308",
    "parent_id": "0000000000000000",
    "name": "http.request",
    "resource": "POST",
    "error": 0,
    "meta": {
        "_dd.p.dm": "-0",
        "_dd.git.commit.sha": "<REDACTED HASH>",
        "_dd.git.repository_url": "https://bitbucket.org/<REDACTED ORG>/<REDACTED NAME>",
        "git.commit.sha": "<REDACTED HASH>",
        "git.repository_url": "https://bitbucket.org/<REDACTED ORG>/<REDACTED NAME>",
        "_dd.origin": "lambda",
        "service": "<REDACTED NAME>",
        "env": "stage",
        "version": "106",
        "runtime-id": "1e55e618-c5af-489b-a469-4e21c624531f",
        "component": "http",
        "span.kind": "client",
        "http.method": "POST",
        "http.url": "http://127.0.0.1:9001/2018-06-01/runtime/init/error",
        "out.host": "127.0.0.1",
        "http.status_code": "202",
        "language": "javascript"
    },
    "metrics": {
        "_dd.agent_psr": 1,
        "_dd.top_level": 1,
        "_dd.measured": 1,
        "network.destination.port": 9001,
        "process_id": 16,
        "_sampling_priority_v1": 1
    },
    "start": 1690819040481031200,
    "duration": 17770264,
    "service": "<REDACTED NAME>",
    "type": "http"
}
,
{
    "trace_id": "4150753a5afad308",
    "span_id": "3dc83454ce76e8a6",
    "parent_id": "4150753a5afad308",
    "name": "tcp.connect",
    "resource": "127.0.0.1:9001",
    "error": 0,
    "meta": {
        "git.commit.sha": "<REDACTED HASH>",
        "git.repository_url": "https://bitbucket.org/<REDACTED ORG>/<REDACTED NAME>",
        "_dd.origin": "lambda",
        "service": "<REDACTED NAME>",
        "env": "stage",
        "version": "106",
        "runtime-id": "1e55e618-c5af-489b-a469-4e21c624531f",
        "component": "net",
        "span.kind": "client",
        "tcp.remote.host": "127.0.0.1",
        "tcp.family": "IPv4",
        "tcp.local.address": "127.0.0.1",
        "out.host": "127.0.0.1",
        "language": "javascript"
    },
    "metrics": {
        "_dd.measured": 1,
        "tcp.remote.port": 9001,
        "tcp.local.port": 34318,
        "network.destination.port": 9001,
        "process_id": 16,
        "_sampling_priority_v1": 1
    },
    "start": 1690819040486973000,
    "duration": 6818604,
    "service": "<REDACTED NAME>"
}
]
2023-07-31T15:57:20.501Z	undefined	DEBUG	Request to the agent: 
{
    "path": "/v0.4/traces",
    "method": "PUT",
    "headers": {
        "Content-Type": "application/msgpack",
        "Datadog-Meta-Tracer-Version": "3.29.1",
        "X-Datadog-Trace-Count": "1",
        "Datadog-Meta-Lang": "nodejs",
        "Datadog-Meta-Lang-Version": "v16.20.1",
        "Datadog-Meta-Lang-Interpreter": "v8"
    },
    "url": "http://127.0.0.1:8126/"
}

EXTENSION	Name: datadog-agent	State: Ready	Events: [INVOKE, SHUTDOWN]
```

**Comment by kcveitch** [comment]
at node_modules/aws-sdk/lib/sequential_executor.js (/var/task/app.js:3154:41)

MAPS TO LINE 5:

```

// node_modules/aws-sdk/lib/sequential_executor.js
var require_sequential_executor = __commonJS({
  "node_modules/aws-sdk/lib/sequential_executor.js"(exports, module2) {
    var AWS3 = require_core2();
    AWS3.SequentialExecutor = AWS3.util.inherit({
      constructor: function SequentialExecutor() {
        this._events = {};
      },
      /**
       * @api private
       */
      listeners: function listeners(eventName) {
        return this._events[eventName] ? this._events[eventName].slice(0) : [];
      },
      on: function on(eventName, listener, toHead) {
        if (this._events[eventName]) {
          toHead ? this._events[eventName].unshift(listener) : this._events[eventName].push(listener);
        } else {
          this._events[eventName] = [listener];
        }
        return this;
      },
      onAsync: function onAsync(eventName, listener, toHead) {
        listener._isAsync = true;
        return this.on(eventName, listener, toHead);
      },
      removeListener: function removeListener(eventName, listener) {
        var listeners = this._events[eventName];
        if (listeners) {
          var length = listeners.length;
          var position = -1;
          for (var i = 0; i < length; ++i) {
            if (listeners[i] === listener) {
              position = i;
            }
          }
          if (position > -1) {
            listeners.splice(position, 1);
          }
        }
        return this;
      },
      removeAllListeners: function removeAllListeners(eventName) {
        if (eventName) {
          delete this._events[eventName];
        } else {
          this._events = {};
        }
        return this;
      },
      /**
       * @api private
       */
      emit: function emit(eventName, eventArgs, doneCallback) {
        if (!doneCallback)
          doneCallback = function() {
          };
        var listeners = this.listeners(eventName);
        var count = listeners.length;
        this.callListeners(listeners, eventArgs, doneCallback);
        return count > 0;
      },
      /**
       * @api private
       */
      callListeners: function callListeners(listeners, args, doneCallback, prevError) {
        var self2 = this;
        var error = prevError || null;
        function callNextListener(err) {
          if (err) {
            error = AWS3.util.error(error || new Error(), err);
            if (self2._haltHandlersOnError) {
              return doneCallback.call(self2, error);
            }
          }
          self2.callListeners(listeners, args, doneCallback, error);
        }
        while (listeners.length > 0) {
          var listener = listeners.shift();
          if (listener._isAsync) {
            listener.apply(self2, args.concat([callNextListener]));
            return;
          } else {
            try {
              listener.apply(self2, args);
            } catch (err) {
              error = AWS3.util.error(error || new Error(), err);
            }
            if (error && self2._haltHandlersOnError) {
              doneCallback.call(self2, error);
              return;
            }
          }
        }
        doneCallback.call(self2, error);
      },
      /**
       * Adds or copies a set of listeners from another list of
       * listeners or SequentialExecutor object.
       *
       * @param listeners [map<String,Array<Function>>, AWS.SequentialExecutor]
       *   a list of events and callbacks, or an event emitter object
       *   containing listeners to add to this emitter object.
       * @return [AWS.SequentialExecutor] the emitter object, for chaining.
       * @example Adding listeners from a map of listeners
       *   emitter.addListeners({
       *     event1: [function() { ... }, function() { ... }],
       *     event2: [function() { ... }]
       *   });
       *   emitter.emit('event1'); // emitter has event1
       *   emitter.emit('event2'); // emitter has event2
       * @example Adding listeners from another emitter object
       *   var emitter1 = new AWS.SequentialExecutor();
       *   emitter1.on('event1', function() { ... });
       *   emitter1.on('event2', function() { ... });
       *   var emitter2 = new AWS.SequentialExecutor();
       *   emitter2.addListeners(emitter1);
       *   emitter2.emit('event1'); // emitter2 has event1
       *   emitter2.emit('event2'); // emitter2 has event2
       */
      addListeners: function addListeners(listeners) {
        var self2 = this;
        if (listeners._events)
          listeners = listeners._events;
        AWS3.util.each(listeners, function(event, callbacks) {
          if (typeof callbacks === "function")
            callbacks = [callbacks];
          AWS3.util.arrayEach(callbacks, function(callback2) {
            self2.on(event, callback2);
          });
        });
        return self2;
      },
      /**
       * Registers an event with {on} and saves the callback handle function
       * as a property on the emitter object using a given `name`.
       *
       * @param name [String] the property name to set on this object containing
       *   the callback function handle so that the listener can be removed in
       *   the future.
       * @param (see on)
       * @return (see on)
       * @example Adding a named listener DATA_CALLBACK
       *   var listener = function() { doSomething(); };
       *   emitter.addNamedListener('DATA_CALLBACK', 'data', listener);
       *
       *   // the following prints: true
       *   console.log(emitter.DATA_CALLBACK == listener);
       */
      addNamedListener: function addNamedListener(name, eventName, callback2, toHead) {
        this[name] = callback2;
        this.addListener(eventName, callback2, toHead);
        return this;
      },
      /**
       * @api private
       */
      addNamedAsyncListener: function addNamedAsyncListener(name, eventName, callback2, toHead) {
        callback2._isAsync = true;
        return this.addNamedListener(name, eventName, callback2, toHead);
      },
      /**
       * Helper method to add a set of named listeners using
       * {addNamedListener}. The callback contains a parameter
       * with a handle to the `addNamedListener` method.
       *
       * @callback callback function(add)
       *   The callback function is called immediately in order to provide
       *   the `add` function to the block. This simplifies the addition of
       *   a large group of named listeners.
       *   @param add [Function] the {addNamedListener} function to call
       *     when registering listeners.
       * @example Adding a set of named listeners
       *   emitter.addNamedListeners(function(add) {
       *     add('DATA_CALLBACK', 'data', function() { ... });
       *     add('OTHER', 'otherEvent', function() { ... });
       *     add('LAST', 'lastEvent', function() { ... });
       *   });
       *
       *   // these properties are now set:
       *   emitter.DATA_CALLBACK;
       *   emitter.OTHER;
       *   emitter.LAST;
       */
      addNamedListeners: function addNamedListeners(callback2) {
        var self2 = this;
        callback2(
          function() {
            self2.addNamedListener.apply(self2, arguments);
          },
          function() {
            self2.addNamedAsyncListener.apply(self2, arguments);
          }
        );
        return this;
      }
    });
    AWS3.SequentialExecutor.prototype.addListener = AWS3.SequentialExecutor.prototype.on;
    module2.exports = AWS3.SequentialExecutor;
  }
});
```

**Comment by tlhunter** [comment]
Thanks, I was typing a request for that content just as you pasted it!

The output is a little surprising to me. Here I can see that the `aws-sdk` package has been included in the bundle even though you've configured esbuild to exclude it. I'll see if I can reproduce this locally.

**Comment by kcveitch** [comment]
well, I got it working.  Pipeline had a extra "npm install" without the --production flag prior to the "sam build" which invokes the esbuild plugin.  So, that was causing it to install the dev dependencies which included the aws-sdk.  I would still expect esbuild to ignore them when packing it up but maybe I'm making too many assumptions there.

after removing the dev dependencies from the build, the lambda executes correctly and looks to be performing without issues.

**Comment by kcveitch** [comment]
awesome.  Thank you so much for helping me work through it.  Glad to hear my pipeline mistakes can lead to potential improvements.

also,  may want to update esbuild example config on: https://docs.datadoghq.com/tracing/trace_collection/dd_libraries/nodejs/?tab=containers#esbuild-support

to include example External modules if using lambda layers:
`
external: ['aws-sdk', 'datadog-lambda-js', 'dd-trace' ],  // If using lambda layers
 `

**Comment by kcveitch** [comment]
On the downside, I was hoping to see the mongoose integration auto-instrumented and It looks like that's still not happening.  But I think that's gonna be a separate ticket.  but FYI for your notes:

2023-07-31T16:29:17.982Z	9542fb78-cf4b-4bbd-a303-0685a06730da	INFO	[dd.trace_id=5821948069728827675 
```

dd.span_id=2600710716659922264] DATADOG TRACER CONFIGURATION - 
{
    "date": "2023-07-31T16:29:17.981Z",
    "os_name": "Linux",
    "os_version": "4.14.255-311-248.529.amzn2.x86_64",
    "architecture": "x64",
    "version": "3.29.1",
    "lang": "nodejs",
    "lang_version": "16.20.1",
    "env": "stage",
    "service": "<REDACTED>",
    "agent_url": "http://127.0.0.1:8126",
    "debug": true,
    "sample_rate": 1,
    "sampling_rules": [],
    "tags": {
        "git.commit.sha": "<REDACTED>",
        "git.repository_url": "<REDACTED>",
        "_dd.origin": "lambda",
        "service": "<REDACTED>",
        "env": "stage",
        "version": "107",
        "runtime-id": "27f9e60f-61af-44fc-adb5-e3e373a96932"
    },
    "dd_version": "107",
    "log_injection_enabled": true,
    "runtime_metrics_enabled": false,
    "profiling_enabled": true,
    "integrations_loaded": [
        "net",
        "dns",
        "http",
        "fetch",
        "aws-sdk"
    ],
    "appsec_enabled": false
}

```

**Comment by tlhunter** [comment]
It looks like the Mongoose issue is a separate one. I'll close this issue out since your original problem is solved. Feel free to open a separate issue for Mongoose.

Note that a commit did land that fixes issues with `external` modules being included in a build. It was based on research done for this issue.

**Comment by naataaniitsosie** [comment]
@tlhunter 
I am experiencing the same issue. However, my usage is a bit different. My code is an express service deployed to Amazon ECS. I'd prefer not to mark the aws-sdk as external as that complicates my container build process. When I revert the dd-trace version to _exactly_ 3.19.0, the express service runs as expected.

Do you have any suggestions?

The following is my configuration and sample output from my local development command.

**../../../../esbuild.config**
```
"use strict";
const baseConfig = {
  logLevel: "debug",
  bundle: true,
  minify: false,
  platform: "node",
  target: "node18",
  outdir: "dist",
  sourcemap: true,
};

module.exports = baseConfig;
```
**scripts/bundle.js**
```
#!/usr/bin/env node
const datadogPlugin = require("dd-trace/esbuild");
const config = require("../../../../esbuild.config");

require("esbuild")
  .build({
    ...config,
    entryPoints: ["src/server.js"],
    plugins: [datadogPlugin],
    external: [
      // mark graphql as external due to an open issue with the datadog plugin
      'graphql/language/visitor',
      'graphql/language/printer',
      'graphql/utilities'
    ]
  })
  .catch(() => process.exit(1));

```


I received the following error when bundling my code:

```
$ npm run serve

> serve
> SKIP_HEADER_VALIDATION=true npm run _serve


> _serve
> nodemon --watch src/ --exec './scripts/bundle.js && NODE_OPTIONS=--enable-source-maps AWS_REGION=us-east-1 DYNAMODB_ENDPOINT=http://localhost:8000 node dist/server.js'

[nodemon] 2.0.22
[nodemon] to restart at any time, enter `rs`
[nodemon] watching path(s): src/**/*
[nodemon] watching extensions: js,mjs,json
[nodemon] starting `./scripts/bundle.js && NODE_OPTIONS=--enable-source-maps AWS_REGION=us-east-1 DYNAMODB_ENDPOINT=http://localhost:8000 node dist/server.js`
● [DEBUG] This call to "require" will not be bundled because the argument is not a string literal [unsupported-require-call]

    ../../../node_modules/express/lib/view.js:81:13:
      81 │     var fn = require(mod).__express
         ╵              ~~~~~~~

● [DEBUG] Indirect calls to "require" will not be bundled [indirect-require]

    ../../../node_modules/node-gyp-build/index.js:6:91:
      6 │ var runtimeRequire = typeof __webpack_require__ === 'function' ? __non_webpack_require__ : require // eslint-disable-line
        ╵                                                                                            ~~~~~~~

● [DEBUG] Using direct eval with a bundler is not recommended and may cause problems [direct-eval]

    ../../../node_modules/@protobufjs/inquire/index.js:12:18:
      12 │         var mod = eval("quire".replace(/^/,"re"))(moduleName); // eslint-disable-line no-eval
         ╵                   ~~~~

  You can read more about direct eval and bundling here: https://esbuild.github.io/link/direct-eval

● [DEBUG] Indirect calls to "require" will not be bundled [indirect-require]

    ../../../node_modules/@aws-sdk/util-user-agent-node/dist-cjs/is-crt-available.js:6:19:
      6 │         if (typeof require === "function" && typeof module !== "undefined" && require("aws-crt")) {
        ╵                    ~~~~~~~

● [DEBUG] This call to "require" will not be bundled because the argument is not a string literal [unsupported-require-call]

    ../../../node_modules/aws-crt/dist/native/binding.js:92:18:
      92 │         binding = require(path);
         ╵                   ~~~~~~~


  dist/server.js      25.8mb ⚠️
  dist/server.js.map  40.4mb

⚡ Done in 1126ms
/Users/tsosie/repos/service-playback/node_modules/aws-sdk/lib/sequential_executor.js:16
AWS.SequentialExecutor = AWS.util.inherit({
                                  ^


TypeError: Cannot read properties of undefined (reading 'inherit')
    at ../../../node_modules/aws-sdk/lib/sequential_executor.js (/Users/tsosie/repos/service-playback/node_modules/aws-sdk/lib/sequential_executor.js:16:35)
    at __require (/Users/tsosie/repos/service-playback/packages/apps/api/dist/server.js:11:51)
    at ../../../node_modules/aws-sdk/lib/core.js (/Users/tsosie/repos/service-playback/node_modules/aws-sdk/lib/core.js:78:1)
    at __require (/Users/tsosie/repos/service-playback/packages/apps/api/dist/server.js:11:51)
    at datadog:/Users/tsosie/repos/service-playback/node_modules/aws-sdk/lib/core.js (datadog:/Users/tsosie/repos/service-playback/node_modules/aws-sdk/lib/core.js:4:19)
    at __require (/Users/tsosie/repos/service-playback/packages/apps/api/dist/server.js:11:51)
    at ../../../node_modules/aws-sdk/lib/event-stream/event-message-chunker-stream.js (/Users/tsosie/repos/service-playback/node_modules/aws-sdk/lib/event-stream/event-message-chunker-stream.js:1:12)
    at __require (/Users/tsosie/repos/service-playback/packages/apps/api/dist/server.js:11:51)
    at ../../../node_modules/aws-sdk/lib/event-stream/streaming-create-event-stream.js (/Users/tsosie/repos/service-playback/node_modules/aws-sdk/lib/event-stream/streaming-create-event-stream.js:8:33)
    at __require (/Users/tsosie/repos/service-playback/packages/apps/api/dist/server.js:11:51)

Node.js v18.15.0
[nodemon] app crashed - waiting for file changes before starting..
```

**Comment by tlhunter** [comment]
@naataaniitsosie it's very interesting that exactly 3.19.0 works for you. Notably, the next version 3.20.0 contains nothing related to esbuild or aws-sdk or express. Are you sure it's exactly 3.19.0 that is the last to function as expected?

Can you explain more about your build process? It appears that the code is possibly bundled both with webpack and esbuild? Or maybe just that one package publishes webpack output to npm...

The number of warnings you're receiving is pretty alarming (express, node-gyp-build, @protobufjs/inquire, @aws-sdk/util-user-agent-node, aws-crt). I haven't seen any of those in my testing. The line in @protobufjs/inquire looks like something a hacker would write and seems to be a way to circumvent bundlers by obscuring the require call. If they only do that for internal node.js modules it would be okay, but if they're loading anything external then it would definitely break at runtime.

Does the application compile and run just fine with esbuild when dd-trace isn't involved?

**Comment by naataaniitsosie** [comment]
@tlhunter 

I am using **esbuild**. My application runs just fine without dd-trace.

I did some in-depth testing. The following versions give me the issues I described previously:

- 3.20.0 to 3.34.0
- 4.9.0 to 4.13.0

It's interesting to note that 3.20.0 and 4.9.0 have the following bug fixes on the [release page](https://github.com/DataDog/dd-trace-js/releases?page=2) change log:
- **esbuild**: fix instrumentation of internal modules in supported packages (https://github.com/DataDog/dd-trace-js/pull/3448)

I upgraded to 4.8.1 so at least that is a win 😃 .

**Comment by tlhunter** [comment]
I'm reopening this issue as aws-sdk currently can't be built with the dd-trace esbuild bundler plugin.

**Comment by tlhunter** [comment]
I've reproduced the issue and added a test case in the linked PR.

Currently I'm going through the compiled bundled both with and without the plugin enabled and am diffing the output. So far nothing stands out as being wrong. The changes appear exactly as intended.

<img width="1751" alt="Screenshot 2023-08-31 at 14 28 54" src="https://github.com/DataDog/dd-trace-js/assets/551402/45d0e678-91db-417b-b170-3522fa2c5efc">
